enarx-keep-sev-shim/src/hostcall.rs

```diff
@@ -0,0 +1,231 @@
+    pub fn try_lock() -> Option<Self> {
+        // Try to take out the virtual address out of the global variable
+        let address = unsafe { HOSTCALL_VIRT_ADDR.swap(0, Ordering::Acquire) };
+        if address == 0 {
+            // Already in use
+            None
+        } else {
+            Some(unsafe { Self(&mut *(address as *mut Block)) })
+        }
+    }
+
+    /// Acquire a HostCall instance to communicate with the host.
+    ///
+    /// # Caution
+    ///
+    /// Do not call this in interrupts, as it can cause deadlocks
+    ///
+    /// # Safety
+    ///
+    /// exclusive mutable access is ensured via an Atomic variable
+    #[inline(always)]
+    pub fn lock() -> Self {
+        loop {
+            // Try to take out the virtual address out of the global variable
+            let address = unsafe { HOSTCALL_VIRT_ADDR.swap(0, Ordering::Acquire) };
+            if address == 0 {
+                // Already in use
+                core::sync::atomic::spin_loop_hint();
+            } else {
+                return unsafe { Self(&mut *(address as *mut Block)) };
+            }
+        }
+    }
+
+    /// Causes a `#VMEXIT` for the host to process the data in the shared memory
+    ///
+    /// Returns the contents of the shared memory reply status, the host might have
+    /// written.
+    ///
+    /// # Safety
+    ///
+    /// The parameters returned can't be trusted.
+    #[inline(always)]
+    unsafe fn hostcall(&mut self) -> Result<[Register<usize>; 2], libc::c_int> {
+        let mut port = Port::<u16>::new(SYSCALL_TRIGGER_PORT);
+        port.write(1 as u16);
+        self.0.msg.rep.into()
+    }
+
+    /// Write `bytes` to a host file descriptor `fd`
+    ///
+    /// # Panics
+    ///
+    /// Panics, if `bytes.len()` exceeds `Block::buf_capacity()`
+    ///
+    /// # Safety
+    ///
+    /// The parameters returned can't be trusted.
+    pub unsafe fn write(&mut self, fd: usize, bytes: &[u8]) -> Result<libc::c_int, libc::c_int> {
+        let count = bytes.len();
+        assert!(count <= Block::buf_capacity());
+
+        let buf_address = Address::from(self.0.buf.as_mut_ptr());
+        let shim_virt_address = ShimVirtAddr::try_from(buf_address).map_err(|_| libc::EFAULT)?;
+
+        let request = Request::new(
+            Register::from_raw(libc::SYS_write as usize),
+            &[
+                fd.into(),
+                ShimPhysAddr::from(shim_virt_address).into(),
+                count.into(),
+            ],
+        );
+
+        self.0.msg.req = request;
+        self.0.buf[..count].copy_from_slice(&bytes[..count]);
+        self.hostcall().map(|r| r[0].raw() as _)
+    }
+
+    /// Exit the shim with a `status` code
+    ///
+    /// # Panics
+    ///
+    /// Panics, if the shim resumes to run.
+    #[inline(always)]
+    pub fn exit_group(&mut self, status: u32) -> ! {
+        unsafe {
+            let request = Request::new(
+                Register::from_raw(libc::SYS_exit_group as usize),
```

Comment:
```suggestion
                libc::SYS_exit_group,
```

I think the above should work everywhere. If not, then `impl From<T> for Register<usize>` needs a fix.